gpcontrib/arenadata_toolkit/src/track_files.c

```diff
@@ -0,0 +1,1367 @@
+		v_stmt.args = list_make1(&arg);
+		elog(LOG, "[arenadata_toolkit] setting relstorages %s in database %u for tracking", buf.data, dbid);
+	}
+
+	tf_guc_unlock_relstorages_once();
+
+	AlterDatabaseSet(&stmt);
+
+	/* Will set the GUC in caller session only on coordinator */
+	tf_guc_unlock_relstorages_once();
+	SetConfigOption("arenadata_toolkit.tracking_relstorages",
+					buf.len ? buf.data : DEFAULT_TRACKED_REL_STORAGES,
+					PGC_S_DATABASE, PGC_S_DATABASE);
+
+	pfree(buf.data);
+
+	PG_RETURN_BOOL(true);
+}
+
+Datum
+tracking_trigger_initial_snapshot(PG_FUNCTION_ARGS)
+{
+	Oid			dbid = get_dbid(PG_GETARG_OID(0));
+	bloom_op_ctx_t ctx = {0};
+
+	tf_check_shmem_error();
+
+	if (Gp_role != GP_ROLE_DISPATCH && IS_QUERY_DISPATCHER())
+	{
+		ereport(ERROR,
+				(errmsg("Cannot execute tracking_trigger_initial_snapshot outside query dispatcher")));
+	}
+
+	elog(LOG, "[arenadata_toolkit] tracking_trigger_initial_snapshot dbid: %u", dbid);
+
+	ctx = bloom_set_get_entry(MyDatabaseId, LW_SHARED, LW_EXCLUSIVE);
+
+	if (!ctx.entry)
+	{
+		bloom_set_release(&ctx);
+		ereport(ERROR,
+		(errmsg("Failed to find corresponding filter to database %u", dbid)));
+	}
+
+	if (Gp_role == GP_ROLE_DISPATCH && !pg_atomic_unlocked_test_flag(&ctx.entry->capture_in_progress))
+	{
+		bloom_set_release(&ctx);
+		ereport(ERROR,
+		  (errmsg("Cannot modify track during track acquisition %u", dbid)));
+	}
+
+	bloom_set_all(&ctx.entry->bloom);
+	bloom_set_release(&ctx);
+
+	if (Gp_role == GP_ROLE_DISPATCH)
+	{
+		char	   *cmd = psprintf("select arenadata_toolkit.tracking_trigger_initial_snapshot(%u)", dbid);
+
+		CdbDispatchCommand(cmd, 0, NULL);
+	}
+
+	PG_RETURN_BOOL(true);
+}
+
+Datum
+tracking_is_initial_snapshot_triggered(PG_FUNCTION_ARGS)
+{
+	Oid			dbid = get_dbid(PG_GETARG_OID(0));
+	bool		is_triggered = false;
+
+	tf_check_shmem_error();
+
+	is_triggered = bloom_set_is_all_bits_triggered(dbid);
+
+	elog(LOG, "[arenadata_toolkit] is_initial_snapshot_triggered:%d dbid: %u", is_triggered, dbid);
+
+	PG_RETURN_BOOL(is_triggered);
+}
+
+Datum
+tracking_is_segment_initialized(PG_FUNCTION_ARGS)
+{
+	ReturnSetInfo *rsi;
+	TupleDesc	tupdesc;
+	HeapTuple	tuple;
+	Datum		values[2];
+	bool		nulls[2] = {false, false};
+	Datum		result;
+
+	tf_check_shmem_error();
+
+	rsi = (ReturnSetInfo *) fcinfo->resultinfo;
+	tupdesc = rsi->expectedDesc;
+
+	/* Populate an output tuple. */
+	values[0] = Int32GetDatum(GpIdentity.segindex);
+	values[1] = BoolGetDatum(pg_atomic_unlocked_test_flag(&tf_shared_state->tracking_is_initialized) == false);
+
+	tuple = heap_form_tuple(tupdesc, values, nulls);
+	result = HeapTupleGetDatum(tuple);
+
+	PG_RETURN_DATUM(result);
+}
+
+/*
+ * This function should be used as argument for tracking_get_track function to
+ * follow correct transaction semantics. Several calls of the function within
+ * the same transaction return ControlVersion, which says tracking_get_track
+ * to return previous filter state.
+ */
+Datum
+tracking_track_version(PG_FUNCTION_ARGS)
```

Comment:
This function is supposed to be called only on coordinator, correct? Maybe add assert at the beginning of the function that we are on the coordinator?